src/lib/party.ts

```diff
@@ -0,0 +1,784 @@
+import { CircuitValue } from './circuit_value';
+import { Group, Field, Bool, VerificationKey } from '../snarky';
+import { PrivateKey, PublicKey } from './signature';
+import { UInt64, UInt32, Int64 } from './int';
+import * as Mina from './mina';
+
+export type Amount = UInt64;
+export const Amount = UInt64;
+export type Balance = UInt64;
+export const Balance = UInt64;
+export type Fee = UInt64;
+export const Fee = UInt64;
+export type GlobalSlot = UInt32;
+export const GlobalSlot = UInt32;
+export type SignedAmount = Int64;
+export const SignedAmount = Int64;
+
+const SnappStateLength: number = 8;
+
+/**
+ * Timing info inside an account.
+ */
+export class Timing {
+  initialMinimumBalance: Balance;
+  cliffTime: GlobalSlot;
+  cliffAmount: Amount;
+  vestingPeriod: GlobalSlot;
+  vestingIncrement: Amount;
+
+  constructor(
+    initialMinimumBalance: Balance,
+    cliffTime: GlobalSlot,
+    cliffAmount: Amount,
+    vestingPeriod: GlobalSlot,
+    vestingIncrement: Amount
+  ) {
+    this.initialMinimumBalance = initialMinimumBalance;
+    this.cliffTime = cliffTime;
+    this.cliffAmount = cliffAmount;
+    this.vestingPeriod = vestingPeriod;
+    this.vestingIncrement = vestingIncrement;
+  }
+}
+
+/**
+ * Either set a value or keep it the same.
+ */
+export class SetOrKeep<T> {
+  set: Bool;
+  value: T;
+
+  setValue(x: T) {
+    this.set = new Bool(true);
+    this.value = x;
+  }
+
+  constructor(set: Bool, value: T) {
+    this.set = set;
+    this.value = value;
+  }
+}
+
+/**
+ * Group a value with a hash.
+ *
+ * @typeParam T the value
+ * @typeParam H the hash
+ */
+export type WithHash<T, H> = {
+  value: T;
+  hash: H;
+};
+
+/**
+ * One specific permission value.
+ *
+ * A [[ Perm ]] tells one specific permission for our snapp how it should behave
+ * when presented with requested modifications.
+ *
+ * Use static factory methods on this class to use a specific behavior. See
+ * documentation on those methods to learn more.
+ */
+export class Perm {
+  constant: Bool;
+  signatureNecessary: Bool;
+  signatureSufficient: Bool;
```

Comment:
It would be less confusing to me if "signatureSufficient" would instead be called "proofNecessary", and the Bool negated (if that's accurate)

Reply:
What exactly does `constant` mean? I thought whether no one has the permission, but below it's also used for the case where everyone has the permission..